src/cljam/io/vcf/util/validator.clj

```diff
@@ -0,0 +1,124 @@
+(ns cljam.io.vcf.util.validator
+  (:require [cljam.io.vcf.util :as vcf-util]))
+
+(defn- valid-ref? [s]
+  (or (nil? s)
+      (every? #{\A \T \G \C \N
+                \a \t \g \c \n} s)))
+
+(defn- valid-qual? [v] (or (nil? v)
+                           (float? v)
+                           (integer? v)))
+
+(defn- valid-filter? [filt]
+  (or (nil? filt)
+      (= filt :PASS)
+      (and (sequential? filt)
+           (every? keyword? filt))))
+
+(defn- check-base-records [variant meta-included-contig? meta-included-format?]
+  (cond-> nil
+    (some #{\< \> \, \  \tab} (:chr variant))
+    (assoc :chr "must not contain whitespaces commas or angle brackets.")
+
+    (not (meta-included-contig? (:chr variant))) (assoc :chr "must be contained in contig header.")
+    (not (integer? (:pos variant))) (assoc :pos "Position must be Integer.")
+    (not (valid-ref? (:ref variant))) (assoc :ref "not include invalid char.")
+    (and (seq (:alt variant))
+         (not (sequential? (:alt variant)))) (assoc :alt "must be sequence.")
+
+    (and (sequential? (:alt variant))
+         (not (every? #(vcf-util/inspect-allele % (:ref variant))
+                      (:alt variant))))
+    (assoc :alt "Invalid allele.")
+
+    (not (valid-qual? (:qual variant))) (assoc :qual "qual must be float.")
+    (not (valid-filter? (:filter variant))) (assoc :filter "Invalid filter.")
+    (not (meta-included-format? (:format variant)))
+    (assoc :format "must be contain meta")))
+
+(defn- check-entry-type [entry type-str]
+  ((case type-str
+     "Integer" integer?
+     "Float" #(or (float? %) (integer? %))
+     "Character" char?
+     "String" string?
+     "Flag" (constantly true)
+     (constantly false))
+   entry))
+
+(defn- check-entry [entry {:keys [type number]}]
+  (let [entries (if (and (= number 1) (not (sequential? entry)))
+                  [entry] entry)
+        type-check (and entry
+                        (not (every? #(check-entry-type
+                                       % type) entries)))
+        number-check (not (or (= number "A")
```

Comment:
Also validate the number of entries for `A`, `R`, and `G`.
- A: `(count :alt)`
- R: `(inc (count :alt))`
- G: `(count (cljam.io.vcf.util/genotype-seq 2 (count :alt)))` or `(inc (cljam.io.vcf.util/genotype-index [(count :alt) (count :alt)]))`